include/revng/Model/IRHelpers.h

```diff
@@ -58,3 +58,55 @@ inline model::TypePath createEmptyStruct(model::Binary &Binary, uint64_t Size) {
   NewStruct->Size = Size;
   return Path;
 }
+
+struct DataObjectSymbol {
+  MetaAddress Address;
+  uint64_t Size;
+  llvm::StringRef Name;
+
+  DataObjectSymbol(MetaAddress Address, uint64_t Size, llvm::StringRef Name) :
+    Address(Address), Size(Size), Name(Name) {}
+
+  bool operator==(const DataObjectSymbol &) const = default;
+};
+
+using DataSymbolVecTy = llvm::SmallVectorImpl<DataObjectSymbol>;
+
+inline void fillSegmentTypeStruct(model::Binary &Binary,
+                                  model::TypePath &Path,
+                                  model::Segment &Segment,
+                                  DataSymbolVecTy &DataSymbols) {
+  auto *Struct = llvm::cast<model::StructType>(Path.get());
+
+  for (const auto &DataObjectSymbol : DataSymbols) {
+    const auto &[Address, Size, Name] = DataObjectSymbol;
+
+    if (Segment.contains(Address)) {
+      auto Offset = Address - Segment.StartAddress;
+      revng_assert(Offset.has_value());
+
+      model::QualifiedType FieldType;
+      model::TypePath T;
+
+      if (Size == 2)
+        T = Binary.getPrimitiveType(model::PrimitiveTypeKind::Unsigned, 2);
+      else if (Size == 4)
+        T = Binary.getPrimitiveType(model::PrimitiveTypeKind::Unsigned, 4);
+      else if (Size == 8)
+        T = Binary.getPrimitiveType(model::PrimitiveTypeKind::Unsigned, 8);
+      else
+        T = Binary.getPrimitiveType(model::PrimitiveTypeKind::Generic, 1);
+
+      if (Size == 1 || Size == 2 || Size == 4 || Size == 8)
+        FieldType = { T, {} };
+      else
+        FieldType = { T, { { model::QualifierKind::Array, Size } } };
+
+      model::StructField Field{ *Offset, {}, Name.str(), FieldType };
+      const auto &[_, Success] = Struct->Fields.insert(Field);
+      revng_assert(Success);
```

Comment:
Made clearer that the struct contains segments and segments only by creating an empty struct within this function, and returning it populated.

Reply:
In `ELFImporter.h` you're replacing the type of `Segment` unconditionally, this is not OK.
You should create an empty struct if there's no `Type` yet, and introduce new fields in empty spaces otherwise.